services/web/server/src/simcore_service_webserver/login/handlers_registration.py

```diff
@@ -190,26 +187,40 @@ async def register(request: web.Request):
         if invitation.trial_account_days:
             expires_at = datetime.utcnow() + timedelta(invitation.trial_account_days)
 
-    username = _get_user_name(registration.email)
-    user: dict = await db.create_user(
-        {
-            "name": username,
-            "email": registration.email,
-            "password_hash": encrypt_password(registration.password.get_secret_value()),
-            "status": (
+    #  get authorized user or create new
+    user = await _auth_api.get_user_by_email(request.app, email=registration.email)
+    if user:
+        await _auth_api.check_authorized_user_or_raise(
+            user,
+            product=product,
+            email=registration.email,
+            password=registration.password.get_secret_value(),
+        )
+    else:
+        user = await _auth_api.create_user(
+            request.app,
+            email=registration.email,
+            password=registration.password.get_secret_value(),
+            status=(
                 CONFIRMATION_PENDING
                 if settings.LOGIN_REGISTRATION_CONFIRMATION_REQUIRED
                 else ACTIVE
             ),
-            "role": USER,
-            "expires_at": expires_at,
-        }
+            expires_at=expires_at,
+        )
+
+    # setup user groups
+    assert (  # nosec
```

Comment:
This is a bit confusing to me. I don't see the reason for this one. Since if invitation is None below code should raise an error.

Reply:
There are 3 possible states. 
1. Invitation is not required (i.e. the app has disabled invitations)
2. Invitation is invalid
3. Invitation is valid

For those states the `invitation` variable get the following values
1. `None`
2. no value, it raises and exception
3. gets `InvitationData`

IN addition, for 3, there are two types of invitations:
1. the one produced by the `invitation service` (new type)
4. the ones created by hand in the database  (deprecated)


NOTE: see now why I have to recfator ? :-)

Reply:
this definitely needs to be split. very hard to go though. thanks for the notes